pkgtools/rpm2pkg/files/rpm2pkg.8
.\"	$NetBSD: rpm2pkg.8,v 1.10 2020/08/26 21:09:47 gutteridge Exp $
.\"
.\" Copyright (c) 2001, 2011 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Emmanuel Dreyfus.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 26, 2020
.Dt RPM2PKG 8
.Os
.Sh NAME
.Nm rpm2pkg
.Nd RPM to BSD package conversion helper
.Sh SYNOPSIS
.Nm
.Op Fl d Ar directory
.Op Fl f Ar package_list
.Op Fl i Ar ignored_dir
.Op Fl p Ar prefix
.Op Fl s Ar strip_path_count
.Ar rpm_file ...
.Sh DESCRIPTION
.Nm
helps in building BSD packages from RPM files by unpacking each specified
.Ar rpm_file ,
and by building a
.Ar package_list
according to what was extracted.
The
.Ar package_list
has the format expected by the
.Nx
package system.
.Pp
.Nm
also provides a mechanism that helps in selecting what is actually
extracted from the RPM files.
It is therefore possible to reject files that are of no interest
when building a BSD package.
This is especially useful for files that are to be extracted in
.Pa /var
or
.Pa /etc .
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl d Ar directory
Change working directory to
.Ar directory .
.It Fl f Ar package_list
Produce a
.Nx
package list
.Ar package_list
from what was actually extracted from the RPM file's archives.
.It Fl i Ar ignored_dir
Do not extract files contained in the
.Ar ignored_dir
directory of the RPM file's archives.
.It Fl p Ar prefix
Use
.Ar prefix
when extracting files.
.It Fl s Ar strip_path_count
Strip
.Ar strip_path_count
leading directories in file paths.
.El
.Sh EXAMPLES
.Nm
.Fl d Ar /usr/pkg
.Fl f Ar /usr/pkgsrc/emulator/more_linux_lib/work/PLIST_DYNAMIC
.Fl p Ar emul/linux
.Fl i Ar tmp
.Fl i Ar var
.Fl i Ar usr/tmp
.Ar /usr/pkgsrc/distfiles/more_linux_lib/foo.rpm
.Ar /usr/pkgsrc/distfiles/more_linux_lib/bar.rpm
.Pp
This will extract
.Pa foo.rpm
and
.Pa bar.rpm
into
.Pa /usr/pkg/emul/linux .
A packing list called
.Pa PLIST_DYNAMIC
will be created, with pathnames relative to
.Pa /usr/pkg ,
because this is the directory into which we asked
.Nm
to change with the
.Fl d
option.
.Pp
.Nm
will not extract files contained in the
.Pa tmp ,
.Pa var ,
or
.Pa usr/tmp
directories of the RPM file's archives.
.Pp
.Nm
.Fl s Ar 1
.Fl d Ar /usr/pkg
.Ar /usr/pkgsrc/distfiles/pack/bar.rpm
.Pp
This will extract the
.Ar bar.rpm
file in
.Pa /usr/pkg
while stripping the leading directory of file paths in
.Ar bar.rpm .
This is useful for dealing with RPM archives that contain files like
.Pa usr/bin/foo ,
which you would like to go in
.Pa /usr/pkg/bin .
.Sh COMPATIBILITY
.Nm
should work with any regular RPM file.
.Sh SEE ALSO
.Xr cpio 1 ,
.Xr rpm 8 ,
.Xr rpm2cpio 8
.Sh AUTHORS
.An Matthias Scheler Aq Mt tron@netbsd.org
.br
This man page was written by
.An Emmanuel Dreyfus Aq Mt manu@netbsd.org


doc/NEWS
# $NetBSD: NEWS,v 1.13 2020/03/30 10:04:27 nia Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

lua 5.3 default [nia 2020-03-30]
	The default Lua version was switched to 5.3.
	Since many packages specify a preferred Lua version, this is mainly
	expected to affect users of Lua modules who were using the default
	settings.

gfortran default [nia 2020-03-30]
	gfortran is now the default Fortran compiler instead of g95.
	gfortran is expected to support more pkgsrc platforms and be
	better at compiling newer Fortran code.

python 3.7 default [maya 2019-04-24]
	As Python 2.x versions are nearing end of life, we have switched
	the default to a newer version.
	Python 2.7 packages are still available, and most python packages
	are built for all available python versions.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"
	to invoke this functionality. Set PKG_DESTDIR_SUPPORT to "none"
	to revert to direct installation.

buildlink3.mk layout changed [joerg 2009-03-20]
	The buildlink3.mk files have been reorganized to reduce
	redundant parsing for deeply nested buildlink3 trees.
	The createbuildlink package has been updated to create
	the new format.

PLIST_VARS introduced [jlam 20080413]
	PLIST_VARS is a simpler way to do conditional PLIST entries.
	The old method of doing e.g. PLIST_SUBST+= FOO="@comment " can
	now be done by unconditionally setting PLIST_VARS+=foo and
	conditionally setting PLIST.foo=yes. This will result in substitution
	of ${PLIST.foo} in the PLIST with either "" or "@comment ".

SKIP_PORTABILITY_CHECK is obsolete [rillig 20080220]
	After being deprecated for about 16 months, this early misdesign
	has finally been obsoleted and removed from pkgsrc.

Check for stripped/unstripped binaries [rillig 20080213]
	A new check has been added in mk/check/check-stripped.mk to
	assure that the installed binaries conform to the setting of the
	INSTALL_UNSTRIPPED variable. It is not enabled by default due to
	lack of testing.

OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


pkgtools/rpm2pkg/files/rpm2pkg.c
/*	$NetBSD: rpm2pkg.c,v 1.20.2.1 2011/04/16 21:27:47 spz Exp $	*/

/*-
 * Copyright (c) 2001-2011 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Matthias Scheler.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "fileio.h"
#include "package-list.h"
#include "parse-rpm.h"

/* Structure of a cpio(1) archive. */
#define C_IRUSR			0000400
#define C_IWUSR			0000200
#define C_IXUSR			0000100
#define C_IRGRP			0000040
#define C_IWGRP			0000020
#define C_IXGRP			0000010
#define C_IROTH			0000004
#define C_IWOTH			0000002
#define C_IXOTH			0000001
#define C_ISUID			0004000
#define C_ISGID			0002000
#define C_ISVTX			0001000
#define C_ISDIR			0040000
#define C_ISREG			0100000
#define C_ISCHR			0020000
#define C_ISLNK			0120000

static const unsigned char CPIOMagic[] = {'0','7','0','7','0','1'};

#define CPIO_END_MARKER		"TRAILER!!!"
#define CPIO_FIELD_LENGTH	8

#define CPIO_HDR_INODE		0
#define CPIO_HDR_MODE		1
#define CPIO_HDR_FILESIZE	6
#define CPIO_HDR_NAMESIZE	11
#define CPIO_NUM_HEADERS	13

#define CP_IFMT			0170000

typedef struct ModeMapStruct {
	unsigned long	mm_CPIOMode;
	mode_t		mm_SysMode;
} ModeMap;

ModeMap ModeMapTab[] = {
	{C_IRUSR, S_IRUSR},
	{C_IWUSR, S_IWUSR},
	{C_IXUSR, S_IXUSR},
	{C_IRGRP, S_IRGRP},
	{C_IWGRP, S_IWGRP},
	{C_IXGRP, S_IXGRP},
	{C_IROTH, S_IROTH},
	{C_IWOTH, S_IWOTH},
	{C_IXOTH, S_IXOTH},
	{C_ISUID, S_ISUID},
	{C_ISGID, S_ISGID},
	{C_ISVTX, S_ISVTX},
	{0, 0}
};

static bool
SkipAndAlign(FileHandle *fh, off_t skip)
{
	off_t old_pos, new_pos;
	char buffer[1 << 16];

	old_pos = FileHandleGetPos(fh);
	new_pos = (old_pos + skip + 3) & ~3;
	if (old_pos == new_pos)
		return true;

	while (old_pos < new_pos) {
		off_t length;
		size_t chunk;

		length = new_pos - old_pos;
		chunk = (length > (off_t)sizeof(buffer)) ?
		    (off_t)sizeof(buffer) : length;
		if (!FileHandleRead(fh, buffer, chunk))
			return false;

		old_pos = FileHandleGetPos(fh);
	}

	return true;
}

static char *
StrCat(const char *prefix, const char *suffix)
{
	size_t	prefix_length;
	char	*str;

	prefix_length = strlen(prefix);
	if ((str = malloc(prefix_length + strlen(suffix) + 1)) == NULL) {
	     perror("malloc");
	     exit(EXIT_FAILURE);
	}

	(void)memcpy(str, prefix, prefix_length);
	(void)strcpy(&str[prefix_length], suffix);

	return str;
}

static char **
ArrayAdd(char **array, char *string)
{
	size_t old;

	old = 0;
	if (array != NULL) {
		while (array[old] != NULL)
			old++;
	}

	if ((array = realloc(array, sizeof(char *) * (old + 2))) == NULL) {
		perror("realloc");
		exit(EXIT_FAILURE);
	}

	array[old++] = string;
	array[old] = NULL;

	return array;
}

static void
Usage(char *Progname)
{
	(void)fprintf(stderr,
	    "Usage: %s [-d directory] [-f packlist] [[-i ignorepath] ...]\n"
	    "               [-p prefix] [-s stripcount] rpmfile [...]\n",
	    Progname);
	exit(EXIT_FAILURE);
}

static char *
GetData(FileHandle *In, unsigned long Length)
{
	char *Ptr;

	if ((Ptr = malloc(Length + 1)) != NULL) {
		if (FileHandleRead(In, Ptr, Length) && SkipAndAlign(In, 0)) {
			Ptr[Length] = '\0';
			return Ptr;
		}
		free(Ptr);
	}

	return NULL;
}

static bool
GetCPIOHeader(FileHandle *in, unsigned long *fields, char **name)
{
	char		buffer[CPIO_NUM_HEADERS][CPIO_FIELD_LENGTH];
	char		header[CPIO_FIELD_LENGTH + 1];
	size_t		i, j;
	unsigned long	namelen;

	*name = NULL;

	if (!FileHandleRead(in, buffer, sizeof(CPIOMagic)))
		return false;
	if (memcmp(buffer, CPIOMagic, sizeof(CPIOMagic)) != 0)
		return false;

	if (!FileHandleRead(in, buffer, sizeof(buffer)))
		return false;

	header[CPIO_FIELD_LENGTH] = '\0';
	for (i = 0; i < CPIO_NUM_HEADERS; i++) {
		for (j = 0; j < CPIO_FIELD_LENGTH; j++) {
			if (!isxdigit((unsigned char)buffer[i][j]))
				return false;
			header[j] = buffer[i][j];
		}

		fields[i] = strtoul(header, NULL, 16);
	}

	namelen = fields[CPIO_HDR_NAMESIZE];
	if ((*name = GetData(in, namelen)) == NULL)
		return false;
	return ((*name)[namelen - 1] == '\0');
}

static mode_t
ConvertMode(unsigned long cpio_mode)
{
	mode_t	mode;
	ModeMap	*ptr;

	for (ptr = ModeMapTab, mode = 0; ptr->mm_CPIOMode != 0; ptr++) {
		if (cpio_mode & ptr->mm_CPIOMode)
			mode |= ptr->mm_SysMode;
	}

	return mode;
}

static bool
MakeTargetDir(char *Name, PListEntry **Dirs)
{
	char		*Basename;
	PListEntry	*Dir;
	struct stat	Stat;
	int	Result;

	if ((Basename = strrchr(Name, '/')) == NULL)
		return true;

	*Basename = '\0';
	if ((Dir = PListFind(*Dirs, Name)) != NULL) {
		*Basename = '/';
		Dir->pe_DirEmpty = false;
		return true;
	}

	if (!MakeTargetDir(Name, Dirs)) {
		*Basename = '/';
		return false;
	}

	if (stat(Name, &Stat) == 0) {
		Result = S_ISDIR(Stat.st_mode);
	} else if (errno != ENOENT) {
		Result = false;
	} else if ((Result = (mkdir(Name, S_IRWXU|S_IRWXG|S_IRWXO) == 0))) {
		(void)PListInsert(Dirs, Name);
	}

	*Basename = '/';
	return Result;
}

static bool
MakeDir(char *name, mode_t mode, bool *old_dir)
{
	struct stat sb;

	*old_dir = false;
	if (mkdir(name, mode) == 0)
		return true;

	if (errno != EEXIST || lstat(name, &sb) < 0 || !S_ISDIR(sb.st_mode))
		return false;

	*old_dir = true;
	return true;
}

static bool
MakeSymLink(char *Link, char *Name)
{
	struct stat Stat;

	if (symlink(Link, Name) == 0) return true;

	if ((errno != EEXIST) || (lstat(Name, &Stat) < 0) ||
	    !S_ISLNK(Stat.st_mode)) {
		return false;
	}

	return ((unlink(Name) == 0) && (symlink(Link, Name) == 0));
}

static bool
WriteFile(FileHandle *in, char *name, mode_t mode, unsigned long length,
    const char *link_target)
{
	int		outfd;
	struct stat	sb;
	char		buffer[1 << 16];

	if ((lstat(name, &sb) == 0) &&
	    (!S_ISREG(sb.st_mode) || (unlink(name) < 0))) {
		return false;
	}

	if (link_target != NULL) {
		if (link(link_target, name) < 0)
			return false;
		outfd = open(name, O_WRONLY, mode);
	 } else {
		outfd = open(name, O_WRONLY|O_CREAT, mode);
	}
	if (outfd < 0)
		return false;

	while (length > 0) {
		ssize_t	chunk;

		chunk = (length > sizeof(buffer)) ? sizeof(buffer) : length;
		if (!FileHandleRead(in, buffer, chunk) ||
		    write(outfd, buffer, chunk) != chunk)
			break;
		length -= chunk;
	}

	if (close(outfd) == 0 && length == 0)
		return SkipAndAlign(in, 0);

	(void)unlink(name);
	return false;
}

static bool
CheckPrefix(const char *prefix, char *name)
{
	size_t length;

	length = strlen(prefix);
	return ((strncmp(prefix, name, length) == 0) &&
	    ((name[length] == '\0') || (name[length] == '/')));
}

static bool
StripPrefix(char *name, int stripcount)
{
	char *new_name;

	if (stripcount <= 0)
		return true;

	new_name = name;
	while (stripcount-- > 0) {
		new_name = strchr(new_name, '/');
		if (new_name == NULL)
			return false;
		new_name++;
	}
	(void)memmove(name, new_name, strlen(new_name) + 1);

	return true;
}

static void
ProcessRPM(const char *filename, PListEntry **files, PListEntry **dirs,
    char **ignore, const char *prefix, int stripcount)
{
	int		fd;
	FileHandle	*in;
	PListEntry	*last;

	if ((fd = open(filename, O_RDONLY, 0)) < 0) {
		perror(filename);
		exit(EXIT_FAILURE);
	}

	if (!IsRPMFile(fd)) {
		(void)fprintf(stderr, "%s: file is not an RPM package.\n",
		    filename);
		exit(EXIT_FAILURE);
	}

	if ((in = OpenRPM(&fd)) == NULL) {
		(void)fprintf(stderr, "%s: cannot get RPM data.\n", filename);
		exit(EXIT_FAILURE);
	}
	if (fd >= 0) {
		(void)close(fd);
		fd = -1;
	}

	last = NULL;
	for (;;) {
		unsigned long	fields[CPIO_NUM_HEADERS];
		char		*name;
		mode_t		mode;
		unsigned long	length;

		if (!GetCPIOHeader(in, fields, &name)) {
			(void)fprintf(stderr,
			    "%s: error in cpio header.\n",
			    filename);
			exit(EXIT_FAILURE);
		}
		if (strcmp(name, CPIO_END_MARKER) == 0) {
			free(name);
			break;
		}
		if (*name == '\0')
			fields[CPIO_HDR_MODE] = 0;

		if (ignore != NULL) {
			char **ptr;

			for (ptr = ignore; *ptr != NULL; ptr++) {
				if (CheckPrefix(*ptr, name)) {
					fields[CPIO_HDR_MODE] = 0;
					break;
				}
			}
		}

		if (fields[CPIO_HDR_MODE] != 0 &&
		    !StripPrefix(name, stripcount)) {
			(void)fprintf(stderr,
				    "%s: Leading path to strip too "
				    "big (-s %d)\n",
				    filename, stripcount);
			exit(EXIT_FAILURE);
		}

		if (prefix != NULL) {
			char *fullname;

		  	fullname = StrCat(prefix, name);
			free(name);
			name = fullname;
		}

		mode = ConvertMode(fields[CPIO_HDR_MODE]);
		length = fields[CPIO_HDR_FILESIZE];
		switch (fields[CPIO_HDR_MODE] & CP_IFMT) {
		case C_ISDIR: {
			PListEntry	*dir;
			bool		old_dir;

			if (length != 0) {
				(void)fprintf(stderr,
				    "%s: error in cpio file.\n",
				filename);
				exit(EXIT_FAILURE);
			}

			if (!MakeTargetDir(name, dirs)) {
				(void)fprintf(stderr,
				    "%s: can't create parent "
				    "directories for \"%s\".\n",
		                    filename, name);
				exit(EXIT_FAILURE);
			}

			if (!MakeDir(name, mode, &old_dir)) {
				(void)fprintf(stderr,
				    "%s: can't create directory "
				    "\"%s\".\n", filename, name);
				exit(EXIT_FAILURE);
			}

			if (!old_dir) {
				dir = PListInsert(dirs, name);
				dir->pe_DirEmpty = true;
			}
			break;
		}
		case C_ISLNK: {
			char *link_target;

			if ((link_target = GetData(in, length)) == NULL) {
				(void)fprintf(stderr,
				    "%s: error in cpio file.\n",
				filename);
				exit(EXIT_FAILURE);
			}

			if (!MakeTargetDir(name, dirs)) {
				(void)fprintf(stderr,
				    "%s: can't create parent "
				    "directories for \"%s\".\n",
		                    filename, name);
				exit(EXIT_FAILURE);
			}

			if (*link_target == '/') {
				char *ptr;

				(void)memmove(link_target, link_target + 1,
				    strlen(link_target));
				ptr = name;
				if (prefix != NULL)
					ptr += strlen(prefix);

				while ((ptr = strchr(ptr, '/')) != NULL) {
					char *new_link_target;

				    	new_link_target = StrCat("../",
					    link_target);
					free(link_target);
					link_target = new_link_target;
					ptr++;
				}
			}

			if (!MakeSymLink(link_target, name)) {
				(void)fprintf(stderr,
				    "%s: can't create symbolic link "
				    "\"%s\".\n", filename, name);
				exit(EXIT_FAILURE);
			}

	    		PListInsert(files, name)->pe_Link = link_target;
			break;
		}
		case C_ISREG:
			if (!MakeTargetDir(name, dirs)) {
				(void)fprintf(stderr,
				    "%s: can't create parent "
				    "directories for \"%s\".\n",
		                    filename, name);
				exit(EXIT_FAILURE);
			}


			if ((last != NULL) && (last->pe_INode !=
			    fields[CPIO_HDR_INODE])) {
				last = NULL;
			}

			if (!WriteFile(in, name, mode, length,
			    (last != NULL)? last->pe_Name : NULL)) {
				(void)fprintf(stderr,
		                    "%s: can't write file \"%s\".\n",
		                    filename, name);
				exit(EXIT_FAILURE);
			}

			last = PListInsert(files, name);
			last->pe_INode = fields[CPIO_HDR_INODE];
			break;
		default:
			if (length > 0 && !SkipAndAlign(in, length)) {
				(void)fprintf(stderr,
				    "%s: error in cpio file.\n",
				    filename);
				exit(EXIT_FAILURE);
			}
					
		}

		free(name);
	}

	FileHandleClose(in);
}

int
main(int argc, char **argv)
{
	char		*progname;
	FILE		*plist_file;
	char		**ignore, *prefix;
	int		opt, stripcount, i;
	PListEntry	*files, *dirs;

	progname = strrchr(argv[0], '/');
	if (progname == NULL)
		progname = argv[0];
	else
		progname++;

	plist_file = NULL;
	ignore = NULL;
	prefix = NULL;
	stripcount = 0;
	while ((opt = getopt(argc, argv, "d:f:i:p:s:")) != -1) {
		switch (opt) {
		case 'd':
			if (chdir(optarg)) {
				perror(optarg);
				return EXIT_FAILURE;
			}
			break;

		case 'f':
			if (plist_file != NULL)
				(void)fclose(plist_file);
			if ((plist_file = fopen(optarg, "a")) == NULL) {
				perror(optarg);
				return EXIT_FAILURE;
			}
			break;

		case 'i':
			ignore = ArrayAdd(ignore, optarg);
			break;

		case 'p':
			if (strlen(optarg) > 0)
				prefix = optarg;
			break;

		case 's':
			stripcount = atoi(optarg);
			if (stripcount <= 0) {
				(void)fprintf(stderr,
				    "%s: -s argument \"%s\" "
				    "must be a positive integer.\n",
				    progname, optarg);
				return EXIT_FAILURE;
			}
			break;

		default:
			Usage(progname);
		}
	}

	if (argc == optind)
		Usage(progname);

	if (prefix != NULL && prefix[strlen(prefix) - 1] != '/')
		prefix = StrCat(prefix, "/");

	files = NULL;
	dirs = NULL;
	for (i = optind; i < argc ; i++)
		ProcessRPM(argv[i], &files, &dirs, ignore, prefix, stripcount);

	if (plist_file != NULL) {
		PListWrite(files, dirs, plist_file);
		(void)fclose(plist_file);
	}

	return EXIT_SUCCESS;
}
